iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/functional_interface.inl

```diff
@@ -111,7 +113,7 @@ inline Derived& AndThenWithValue<Derived, ValueType>::and_then(const and_then_ca
 {
```

Comment:
This was a discussion we had in the initial functional interface PR.

When we would use the `cxx::string` or `char`-array approach we would have to make this function a template so that it captures the size. You can have a wide variety of messages. We could also decide for a very long string but then we would waste stack memory in most cases.
Furthermore with the `cxx::string` template approach we would have to provide the size since the argument is most likely a string literal and not a `cxx::string`.

The user would have to write:
```cpp
myOptional.expect<11>("hello world"); // somehow ugly
myOptional.expect(someString); // most of the time it will not be used this way but here we could get rid of the template argument
```
-> we concluded that the `cxx::string` are not really fitting our use case.

The char array approach would look like:
```
struct Expect {
  template<uint64_t N>
  void expect( const char[N] msg);
};
```
With this we would cover all the string literal use cases and could write:
```cpp
myOptional.expect("hello world");
```
But now this causes a compile failure:
```cpp
myOptional.expect(someString); 
myOptional.expect(someString.c_str()); // also fails since we require the size as template argument and not a char ptr.
myOptional.expect<11>(stringWhichContainsHelloWorld.c_str()); // this may somehow work but this is evil
```
And think of all code which would be generated by the compiler, for all the length variations of `msg`.
-> so we concluded `char`-array is also not the way to go.

With the major use case in mind, that the user would like to do something like:
```cpp
SharedMemory shm = SharedMemoryBuilder::create().expect("unable to create shared memory");
```
And sometimes:
```cpp
SharedMemory shm = SharedMemoryBuilder::create().expect(myErrorMessage.c_str());
```
we decided then to follow the `const char* const` approach since it covers our major use case perfectly without having an unbearable API or generate a lot of code blaot.

Reply:
Thanks for the detailed description!

There is a question which is still unanswered:

> What happens if the user passes a char array which is not null-terminated?

Let's assume the following

```
char myUnterminatedEvilString[3] = {'f','o','o'};

Foo::create().expect(myUnterminatedEvilString);
// std::cout reads the memory behind myUnterminatedEvilString till a '\0' is found
```

This is a major security problem, because we read the consecutive memory aka out-of-bounds access or am I mistaken here? That exactly the kind of cases where `std::span` should be used in the signature to pass the size alongside the pointer.

Instead of a generic `std::span` we could also make use of `std::string_view` in the signature of `expects(..)`. I know these two are not available in C++14 but I would be happy to backport them to C++14 :blush:

Reply:
I would opt for option 2 in combination with option 1 in order to prevent code bloat.
```
public:
template<uint64_t N>
... expect(const char message[N]) // supports use case 1
{
    return expect(message);
}

template<uint64_t N>
... expect(const cxx::string<N> & message) // supports use case 2
{
    return expect(message.c_str());
}
private:
... expect(const char* message, const uint64_t strSize);
```
There is still monopolization, but it should be optimized out and therefore not increase the binary size.

Reply:
So I choose secret option number 4.
```
template <typename StringContainer>
ValueType& expect(const StringContainer& msg) & noexcept {
  static_assert((std::is_array<StringContainer>::value
               && std::is_same<typename std::remove_all_extents<StringContainer>::type, char>::value)
                  || is_iox_string<StringContainer>::value,
              "Only char arrays and iox::cxx::strings are allowed as message type.");
//...
}
```

I allow every type and restrict it to either a char array or a `cxx::string` in the static assert so the developer gets a nice compiler failure.

I refrained from the overload once with char and once with string since expect already has 4 overloads (lvalue, rvalue and const variations). So adding two additional expect methods once with char array and once with string would produce in the end 12 additional methods.

Reply:
Sadly there is no short way.

It is possible to implement something which is somehow similar to c++20 concepts with c++14 but this requires a lot of more code and the PR is already big enough.

Since the user cannot compile this and gets a sane error message right from the beginning and even my IDE (neovim) realizes that and prints me an error I can say we with good conscience that we are save here!